### ISSUE TITLE
nested partially working and breaking mcp generation

### ISSUE THREAD

**Issue by benoitj** [intellectronica/ruler#248]
I'm running ruler version 0.3.9

I'm testing nested configuration with mcp generation enabled and found two issues:
1. the toml configuration attribute `nested = true` is ignored
2. when enabling --nested, the mcp generation is not triggered, even when enabled in toml, and with --with-mcp argument

A consequence of issue 2 is that it also removes the mcp.json from the gitignore and doing so, makes them candidate to be added to git.

I created a small repo to reproduce the problem: https://github.com/benoitj/ruler-nested

here is the output of apply without parameters, where the root AGENTS.md and the .vscode/mcp.json are created:

```text
> ruler apply --verbose

 [ruler:verbose] Loading configuration from project root: /home/benoit/src/sandbox/ruler-nested
[ruler:verbose] Loaded configuration with 0 agent configs
[ruler:verbose] Found .ruler directory with 376 characters of rules
[ruler:verbose] Selected 1 agents: GitHub Copilot
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /home/benoit/src/sandbox/ruler-nested/AGENTS.md
[ruler:verbose] Applying filtered MCP config for GitHub Copilot with strategy: merge and key: servers
[ruler:verbose] MCP config for GitHub Copilot is already up to date - skipping backup and write
[ruler] Updated .gitignore with 4 unique path(s) in the Ruler block.
Ruler apply completed successfully.
```

here is when passing --nested, and forcing mcp too.
In this case, it creates both the root AGENTS.md and the some-module/AGENTS.md, but does not create the .vscode/mcp.json

```text
> ruler apply --verbose --nested --with-mcp

 [ruler:verbose] Loading configuration from project root: /home/benoit/src/sandbox/ruler-nested
[ruler:verbose] Loaded 2 .ruler directory configurations
[ruler:verbose] Root configuration has 0 agent configs
[ruler:verbose] Selected 1 agents: GitHub Copilot
[ruler] Processing .ruler directory: /home/benoit/src/sandbox/ruler-nested/some-module/.ruler
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /home/benoit/src/sandbox/ruler-nested/some-module/AGENTS.md
[ruler] Processing .ruler directory: /home/benoit/src/sandbox/ruler-nested/.ruler
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /home/benoit/src/sandbox/ruler-nested/AGENTS.md
[ruler] Updated .gitignore with 4 unique path(s) in the Ruler block.
Ruler apply completed successfully.
```

let me know if you need more information.

**Comment by intellectronica** [comment]
Reproduction
- `tmp/issue-248-repro.sh` builds a fresh workspace under `tmp/issue-248`, scaffolds root `.ruler` config + nested `.ruler` folders (`apps/service-a/.ruler`, `libs/shared/.ruler`), and installs `@intellectronica/ruler@0.3.10` from npm into an isolated prefix.
- Scenario 1 runs `node <cli>/dist/cli/index.js apply --verbose` without the `--nested` flag. Scenario 2 reruns the setup and executes `apply --verbose --nested --with-mcp`.

Findings
- `nested = true` in `ruler.toml` is parsed (`src/core/ConfigLoader.ts:205-222`), but `applyAllAgentConfigs` only switches to hierarchical mode when the CLI flag is set (`src/lib.ts:55-96`). Flat runs therefore ignore nested folders, matching scenario 1 where only the root `AGENTS.md` and `.vscode/mcp.json` are produced.
- Hierarchical mode drops the MCP payload. `loadNestedConfigurations` delegates to `createHierarchicalConfiguration`, which hard-codes `rulerMcpJson: null` (`src/core/apply-engine.ts:98-112`). Later, `handleMcpConfiguration` bails out whenever `rulerMcpJson` is falsy (`src/core/apply-engine.ts:433-460`), so `.vscode/mcp.json` is never written and its `.gitignore` entry disappears—exactly what scenario 2 shows.

How to fix
1. Honour the TOML flag by letting the CLI default `nested` to `loadedConfig.nested` (or by moving the toggle inside `applyAllAgentConfigs` so config-sourced values take effect when the flag is omitted).
2. When building hierarchical configs, synthesize the MCP bundle the same way `loadSingleConfiguration` does (`src/core/apply-engine.ts:207-223`) so `handleMcpConfiguration` still executes.

Transcript
```
Using ruler version: 0.3.10
working directory: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248

Installing CLI once for all scenarios...
CLI installed at /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/ruler-cli/node_modules/@intellectronica/ruler/dist/cli/index.js

== Scenario 1: default apply (expect nested config ignored) ==

Root .ruler tree:
AGENTS.md
additional.md
ruler.toml

Nested directories discovered:
./.ruler
./libs/shared/.ruler
./apps/service-a/.ruler

-- running ruler apply --verbose --
[ruler:verbose] Loading configuration from project root: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario1
[ruler:verbose] Loaded configuration with 0 agent configs
[ruler:verbose] Found .ruler directory with 193 characters of rules
[ruler:verbose] Selected 1 agents: GitHub Copilot
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario1/AGENTS.md
[ruler:verbose] Applying filtered MCP config for GitHub Copilot with strategy: merge and key: servers
[ruler] Updated .gitignore with 4 unique path(s) in the Ruler block.
Ruler apply completed successfully.

-- git status after scenario 1 --
## issue-248-analysis
 M ../../../package-lock.json
?? ../../../.specstory/

File checks after scenario 1:
  ✅ root AGENTS.md present (AGENTS.md)
  ❌ apps/service-a AGENTS.md missing (apps/service-a/AGENTS.md)
  ❌ libs/shared AGENTS.md missing (libs/shared/AGENTS.md)
  ✅ root MCP config present (.vscode/mcp.json)

-- .gitignore (Ruler block) --
# START Ruler Generated Files
/.vscode/mcp.json
/.vscode/mcp.json.bak
/AGENTS.md
/AGENTS.md.bak
# END Ruler Generated Files

== Scenario 2: apply with --nested --with-mcp ==

Root .ruler tree:
AGENTS.md
additional.md
ruler.toml

Nested directories discovered:
./.ruler
./libs/shared/.ruler
./apps/service-a/.ruler

-- running ruler apply --verbose --nested --with-mcp --
[ruler:verbose] Loading configuration from project root: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2
[ruler:verbose] Loaded 4 .ruler directory configurations
[ruler:verbose] Root configuration has 0 agent configs
[ruler:verbose] Selected 1 agents: GitHub Copilot
[ruler] Processing .ruler directory: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2/apps/service-a/.ruler
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2/apps/service-a/AGENTS.md
[ruler] Processing .ruler directory: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2/libs/shared/.ruler
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2/libs/shared/AGENTS.md
[ruler] Processing .ruler directory: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2/.ruler
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /Users/eleanor/Library/CloudStorage/GoogleDrive-elle.fnord@gmail.com/My Drive/Repositories/ruler/tmp/issue-248/scenario2/AGENTS.md
[ruler] Processing .ruler directory: /Users/eleanor/.config/ruler
[ruler] Applying rules for GitHub Copilot...
[ruler:verbose] Processing agent: GitHub Copilot
[ruler:verbose] Agent GitHub Copilot output paths: /Users/eleanor/.config/AGENTS.md
[ruler] Updated .gitignore with 8 unique path(s) in the Ruler block.
Ruler apply completed successfully.

-- git status after scenario 2 --
## issue-248-analysis
 M ../../../package-lock.json
?? ../../../.specstory/

File checks after scenario 2:
  ✅ root AGENTS.md present (AGENTS.md)
  ✅ apps/service-a AGENTS.md present (apps/service-a/AGENTS.md)
  ✅ libs/shared AGENTS.md present (libs/shared/AGENTS.md)
  ❌ root MCP config missing (.vscode/mcp.json)

-- .gitignore (Ruler block) --
# START Ruler Generated Files
/../../../../../../../../../.config/AGENTS.md
/../../../../../../../../../.config/AGENTS.md.bak
/AGENTS.md
/AGENTS.md.bak
/apps/service-a/AGENTS.md
/apps/service-a/AGENTS.md.bak
/libs/shared/AGENTS.md
/libs/shared/AGENTS.md.bak
# END Ruler Generated Files
```

**Comment by intellectronica** [comment]
A few clarifications / hints for fixing:

The behaviour should be:
- Nested processing should be triggered either by specifying `--nested` on the CLI or by including a `nested = true` attribute in the toml file.
- The nested behaviour is always determined by the current closest ruler config and applies to subdirectories below it.
- Generation in nested directories should work exactly the same as the top-level directory. It should generate rules and MCP configs in the nested subdirectory. This behaviour doesn't always make sense, but we should leave it up to users to decide how to structure their repositories and configs.
- Configs for nested directories always inherit `nested = true`. Any other behaviour would be too weird. If `nested = false` is specified, we should warn that it is being overridden.
- We should always warn that nested processing is an _experimenal_ feature (until further decision).
- Documentation should be updated to clarify how nested processing works, including some examples and suggestions for how to use it.

**Comment by intellectronica** [comment]
I ran the new `tmp/248-post-fix-repro.sh` helper to sanity-check the nested MCP propagation fix.

Key observations:

- The script scaffolds a 3-level tree (15 `.ruler` directories) with a mix of rule-only, config-only, and MCP-enabled combinations, then drives `npx --yes github:intellectronica/ruler apply --local-only --with-mcp` from a clean temp workspace.
- Nested mode now auto-enables from TOML: even without passing `--nested`, the run emits the expected experimental warning and processes all 15 directories. Configs that set `nested = false` surface the override warning but are still processed.
- MCP bundles are generated per directory wherever configured. The verification table below shows the detected server keys for each location (blank entries are directories where MCP was intentionally disabled):

```
Directory                                          | AGENTS    | MCP       | MCP Servers
.                                                  | present   | present   | root_stdio
team-alpha                                         | present   | present   | team-alpha
team-beta                                          | skip      | present   | team-beta
team-alpha/service-api                             | present   | skip      | n/a
team-alpha/service-web                             | present   | skip      | n/a
team-beta/component-shared                         | present   | present   | team-beta-component-shared
team-beta/component-utils                          | present   | skip      | n/a
team-alpha/service-api/module-db                   | skip      | present   | team-alpha-service-api-module-db
team-alpha/service-api/module-workers              | present   | present   | team-alpha-service-api-module-workers
team-alpha/service-web/module-ui                   | present   | skip      | n/a
team-alpha/service-web/module-admin                | present   | skip      | n/a
team-beta/component-shared/submodule-core          | present   | present   | team-beta-component-shared-submodule-core
team-beta/component-shared/submodule-ops           | skip      | present   | team-beta-component-shared-submodule-ops
team-beta/component-utils/submodule-linters        | present   | skip      | n/a
team-beta/component-utils/submodule-build          | present   | skip      | n/a
```

- `.gitignore` picked up all 46 generated targets (AGENTS + MCP + backups) including the nested paths, matching the expectations from the issue description.

Full transcript (314 lines) lives at `tmp/248-post-fix-repro-output.txt` alongside the script so anyone can re-run with `KEEP_TMP=1` to inspect the workspace.
